File: java-fit-performer/src/main/java/com/couchbase/manager/CollectionManagerHelper.java
/*
 * Copyright 2023 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.manager;

// [skip:<3.4.12]

import com.couchbase.client.core.cnc.RequestSpan;
import com.couchbase.client.java.Cluster;
import com.couchbase.client.java.ReactiveCluster;
import com.couchbase.client.protocol.run.Result;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.CollectionSpec;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.CreateCollectionOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.CreateCollectionSettings;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.CreateScopeOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.DropCollectionOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.DropScopeOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.GetAllScopesOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.GetAllScopesResult;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.ScopeSpec;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.UpdateCollectionOptions;
import com.couchbase.client.protocol.sdk.bucket.collectionmanager.UpdateCollectionSettings;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import static com.couchbase.JavaSdkCommandExecutor.setSuccess;


public class CollectionManagerHelper {

  private CollectionManagerHelper() {

  }

  public static void handleCollectionManager(Cluster cluster,
                                             ConcurrentHashMap<String, RequestSpan> spans,
                                             com.couchbase.client.protocol.sdk.Command command,
                                             Result.Builder result) {

    var cm = command.getBucketCommand().getCollectionManager();
    var bucketName = command.getBucketCommand().getBucketName();
    var collections = cluster.bucket(bucketName).collections();
    if (cm.hasGetAllScopes()) {
      var request = cm.getGetAllScopes();
      List<com.couchbase.client.java.manager.collection.ScopeSpec> response;
      if (!request.hasOptions()) {
        response = collections.getAllScopes();
      } else {
        var options = createGetAllScopesOptions(request.getOptions(), spans);
        response = collections.getAllScopes(options);
      }
      populateResult(result, response);
    } else if (cm.hasCreateScope()) {
      var request = cm.getCreateScope();
      if (request.hasOptions()) {
        var options = createScopeOptions(request.getOptions(), spans);
        collections.createScope(request.getName(), options);
      } else {
        collections.createScope(request.getName());
      }
      setSuccess(result);
    } else if (cm.hasDropScope()) {
      var request = cm.getDropScope();
      if (request.hasOptions()) {
        var options = dropScopeOptions(request.getOptions(), spans);
        collections.dropScope(request.getName(), options);
      } else {
        collections.dropScope(request.getName());
      }
      setSuccess(result);
    } else if (cm.hasCreateCollection()) {
      var request = cm.getCreateCollection();
      var settings = createCollectionSettings(request.hasSettings(), request.getSettings());
      if (request.hasOptions()) {
        var options = createCollectionOptions(request.getOptions(), spans);
        collections.createCollection(request.getScopeName(), request.getName(), settings, options);
      } else {
        collections.createCollection(request.getScopeName(), request.getName(), settings);
      }
      setSuccess(result);
    } else if (cm.hasUpdateCollection()) {
      var request = cm.getUpdateCollection();
      var settings = updateCollectionSettings(request.hasSettings(), request.getSettings());
      if (request.hasOptions()) {
        var options = updateCollectionOptions(request.getOptions(), spans);
        collections.updateCollection(request.getScopeName(), request.getName(), settings, options);
      } else {
        collections.updateCollection(request.getScopeName(), request.getName(), settings);
      }
      setSuccess(result);
    } else if (cm.hasDropCollection()) {
      var request = cm.getDropCollection();
      if (request.hasOptions()) {
        var options = dropCollectionOptions(request.getOptions(), spans);
        collections.dropCollection(request.getScopeName(), request.getName(), options);
      } else {
        collections.dropCollection(request.getScopeName(), request.getName());
      }
      setSuccess(result);
    } else {
      throw new UnsupportedOperationException("Unknown CollectionManager operation");
    }
  }

  public static Mono<Result> handleCollectionManagerReactive(ReactiveCluster cluster,
                                                             ConcurrentHashMap<String, RequestSpan> spans,
                                                             com.couchbase.client.protocol.sdk.Command command,
                                                             Result.Builder result) {


    var cm = command.getBucketCommand().getCollectionManager();
    var bucketName = command.getBucketCommand().getBucketName();
    var collections = cluster.bucket(bucketName).collections();
    if (cm.hasGetAllScopes()) {
      Flux<com.couchbase.client.java.manager.collection.ScopeSpec> response;
      var request = cm.getGetAllScopes();
      if (!request.hasOptions()) {
        response = cluster.bucket(bucketName).collections().getAllScopes();
      } else {
        var options = createGetAllScopesOptions(request.getOptions(), spans);
        response = cluster.bucket(bucketName).collections().getAllScopes(options);
      }
      Flux<Result> f = response.map(r -> {
        populateResult(result, r);
        return result.build();
      });
      return f.last();
    } else if (cm.hasCreateScope()) {
      var request = cm.getCreateScope();
      Mono<Void> r;
      if (request.hasOptions()) {
        var options = createScopeOptions(request.getOptions(), spans);
        r = collections.createScope(request.getName(), options);
      } else {
        r = collections.createScope(request.getName());
      }
      return r.then(Mono.fromSupplier(() -> {
        setSuccess(result);
        return result.build();
      }));
    } else if (cm.hasDropScope()) {
      var request = cm.getDropScope();
      Mono<Void> r;
      if (request.hasOptions()) {
        var options = dropScopeOptions(request.getOptions(), spans);
        r = collections.dropScope(request.getName(), options);
      } else {
        r = collections.dropScope(request.getName());
      }
      return r.then(Mono.fromSupplier(() -> {
        setSuccess(result);
        return result.build();
      }));
    } else if (cm.hasCreateCollection()) {
      var request = cm.getCreateCollection();
      var settings = createCollectionSettings(request.hasSettings(), request.getSettings());
      Mono<Void> r;
      if (request.hasOptions()) {
        var options = createCollectionOptions(request.getOptions(), spans);
        r = collections.createCollection(request.getScopeName(), request.getName(), settings, options);
      } else {
        r = collections.createCollection(request.getScopeName(), request.getName(), settings);
      }
      return r.then(Mono.fromSupplier(() -> {
        setSuccess(result);
        return result.build();
      }));
    } else if (cm.hasUpdateCollection()) {
      var request = cm.getUpdateCollection();
      var settings = updateCollectionSettings(request.hasSettings(), request.getSettings());
      Mono<Void> r;
      if (request.hasOptions()) {
        var options = updateCollectionOptions(request.getOptions(), spans);
        r = collections.updateCollection(request.getScopeName(), request.getName(), settings, options);
      } else {
        r = collections.updateCollection(request.getScopeName(), request.getName(), settings);
      }
      return r.then(Mono.fromSupplier(() -> {
        setSuccess(result);
        return result.build();
      }));
    } else if (cm.hasDropCollection()) {
      var request = cm.getDropCollection();
      Mono<Void> r;
      if (request.hasOptions()) {
        var options = dropCollectionOptions(request.getOptions(), spans);
        r = collections.dropCollection(request.getScopeName(), request.getName(), options);
      } else {
        r = collections.dropCollection(request.getScopeName(), request.getName());
      }
      return r.then(Mono.fromSupplier(() -> {
        setSuccess(result);
        return result.build();
      }));
    } else {
      return Mono.error(new UnsupportedOperationException(new IllegalArgumentException("Unknown operation")));
    }
  }

  private static CollectionSpec.Builder toFit(com.couchbase.client.java.manager.collection.CollectionSpec spec) {
    CollectionSpec.Builder builder = CollectionSpec.newBuilder()
            .setName(spec.name())
            .setExpirySecs(spec.maxExpiry() != null ? (int) spec.maxExpiry().toSeconds() : 0)
            .setScopeName(spec.scopeName());
    if (spec.history() != null) {
      builder.setHistory(spec.history());
    }
    return builder;
  }

  public static void populateResult(Result.Builder result, List<com.couchbase.client.java.manager.collection.ScopeSpec> response) {
    var builder = GetAllScopesResult.newBuilder();

    var allResult = response.stream()
            .map(scopeSpec -> ScopeSpec.newBuilder()
                    .setName(scopeSpec.name())
                    .addAllCollections(scopeSpec.collections().stream()
                            .map(c -> toFit(c).build())
                            .collect(Collectors.toList()))
                    .build()).collect(Collectors.toList());

    builder.addAllResult(allResult);

    result.setSdk(com.couchbase.client.protocol.sdk.Result.newBuilder()
            .setCollectionManagerResult(com.couchbase.client.protocol.sdk.bucket.collectionmanager.Result.newBuilder()
                    .setGetAllScopesResult(builder)));

  }

  public static void populateResult(Result.Builder result, com.couchbase.client.java.manager.collection.ScopeSpec scopeSpec) {
    var builder = GetAllScopesResult.newBuilder();

    var list = result.getSdkBuilder() != null ?
            result.getSdkBuilder().getCollectionManagerResult().getGetAllScopesResult().getResultList() : null;
    var allResult = ScopeSpec.newBuilder()
            .setName(scopeSpec.name())
            .addAllCollections(scopeSpec.collections().stream()
                    .map(c -> toFit(c).build())
                    .collect(Collectors.toList()))
            .build();

    if (list != null) {
      builder.addAllResult(list);
    }
    builder.addResult(allResult);

    result.setSdk(com.couchbase.client.protocol.sdk.Result.newBuilder()
            .setCollectionManagerResult(com.couchbase.client.protocol.sdk.bucket.collectionmanager.Result.newBuilder()
                    .setGetAllScopesResult(builder)));

  }

  private static com.couchbase.client.java.manager.collection.GetAllScopesOptions createGetAllScopesOptions(GetAllScopesOptions getAllScopesOptions, ConcurrentHashMap<String, RequestSpan> spans) {
    var options = com.couchbase.client.java.manager.collection.GetAllScopesOptions.getAllScopesOptions();

    if (getAllScopesOptions.hasTimeoutMsecs()) options.timeout(java.time.Duration.ofMillis(getAllScopesOptions.getTimeoutMsecs()));
    if (getAllScopesOptions.hasParentSpanId()) options.parentSpan(spans.get(getAllScopesOptions.getParentSpanId()));

    return options;
  }

  private static com.couchbase.client.java.manager.collection.CreateScopeOptions createScopeOptions(CreateScopeOptions options, ConcurrentHashMap<String, RequestSpan> spans) {
    var out = com.couchbase.client.java.manager.collection.CreateScopeOptions.createScopeOptions();

    if (options.hasTimeoutMsecs()) out.timeout(java.time.Duration.ofMillis(options.getTimeoutMsecs()));
    if (options.hasParentSpanId()) out.parentSpan(spans.get(options.getParentSpanId()));

    return out;
  }

  private static com.couchbase.client.java.manager.collection.DropScopeOptions dropScopeOptions(DropScopeOptions options, ConcurrentHashMap<String, RequestSpan> spans) {
    var out = com.couchbase.client.java.manager.collection.DropScopeOptions.dropScopeOptions();

    if (options.hasTimeoutMsecs()) out.timeout(java.time.Duration.ofMillis(options.getTimeoutMsecs()));
    if (options.hasParentSpanId()) out.parentSpan(spans.get(options.getParentSpanId()));

    return out;
  }

  private static com.couchbase.client.java.manager.collection.CreateCollectionOptions createCollectionOptions(CreateCollectionOptions options, ConcurrentHashMap<String, RequestSpan> spans) {
    var out = com.couchbase.client.java.manager.collection.CreateCollectionOptions.createCollectionOptions();

    if (options.hasTimeoutMsecs()) out.timeout(java.time.Duration.ofMillis(options.getTimeoutMsecs()));
    if (options.hasParentSpanId()) out.parentSpan(spans.get(options.getParentSpanId()));

    return out;
  }

  private static com.couchbase.client.java.manager.collection.CreateCollectionSettings createCollectionSettings(boolean hasSettings, CreateCollectionSettings settings) {
    var out = com.couchbase.client.java.manager.collection.CreateCollectionSettings.createCollectionSettings();

    if (hasSettings) {
      if (settings.hasExpirySecs()) out.maxExpiry(java.time.Duration.ofSeconds(settings.getExpirySecs()));
      if (settings.hasHistory()) out.history(settings.getHistory());
    }

    return out;
  }

  private static com.couchbase.client.java.manager.collection.UpdateCollectionOptions updateCollectionOptions(UpdateCollectionOptions options, ConcurrentHashMap<String, RequestSpan> spans) {
    var out = com.couchbase.client.java.manager.collection.UpdateCollectionOptions.updateCollectionOptions();

    if (options.hasTimeoutMsecs()) out.timeout(java.time.Duration.ofMillis(options.getTimeoutMsecs()));
    if (options.hasParentSpanId()) out.parentSpan(spans.get(options.getParentSpanId()));

    return out;
  }

  private static com.couchbase.client.java.manager.collection.UpdateCollectionSettings updateCollectionSettings(boolean hasSettings, UpdateCollectionSettings settings) {
    var out = com.couchbase.client.java.manager.collection.UpdateCollectionSettings.updateCollectionSettings();

    if (hasSettings) {
      if (settings.hasExpirySecs()) out.maxExpiry(java.time.Duration.ofSeconds(settings.getExpirySecs()));
      if (settings.hasHistory()) out.history(settings.getHistory());
    }

    return out;
  }

  private static com.couchbase.client.java.manager.collection.DropCollectionOptions dropCollectionOptions(DropCollectionOptions options, ConcurrentHashMap<String, RequestSpan> spans) {
    var out = com.couchbase.client.java.manager.collection.DropCollectionOptions.dropCollectionOptions();

    if (options.hasTimeoutMsecs()) out.timeout(java.time.Duration.ofMillis(options.getTimeoutMsecs()));
    if (options.hasParentSpanId()) out.parentSpan(spans.get(options.getParentSpanId()));

    return out;
  }

}

File: scala-fit-performer/src/main/scala/com/couchbase/client/performer/scala/manager/CollectionManagerHelper.scala
/*
 * Copyright (c) 2023 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.performer.scala.manager

// [skip:<1.4.11]

import com.couchbase.client.performer.scala.ScalaSdkCommandExecutor.setSuccess
import com.couchbase.client.performer.scala.util.OptionsUtil.{
  DefaultManagementTimeout,
  DefaultRetryStrategy
}
import com.couchbase.client.protocol.run.Result
import com.couchbase.client.protocol.sdk.Command
import com.couchbase.client.scala.Cluster
import com.couchbase.client.scala.manager.collection.{
  CollectionSpec,
  CreateCollectionSettings,
  ScopeSpec,
  UpdateCollectionSettings
}

import java.util.concurrent.TimeUnit
import scala.concurrent.duration
import scala.jdk.CollectionConverters._
import scala.util.{Failure, Success}

object CollectionManagerHelper {

  def handleCollectionManager(cluster: Cluster, command: Command): Result.Builder = {
    val bm     = command.getBucketCommand.getCollectionManager
    val bucket = cluster.bucket(command.getBucketCommand.getBucketName)

    val result = Result.newBuilder()

    if (bm.hasGetAllScopes) {
      val req = bm.getGetAllScopes

      val response = bucket.collections.getAllScopes(
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(scopes) => populateResult(result, scopes)
        case Failure(e)      => throw e
      }
    } else if (bm.hasCreateScope) {
      val req = bm.getCreateScope
      val response = bucket.collections.createScope(
        req.getName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasDropScope) {
      val req = bm.getDropScope
      val response = bucket.collections.dropScope(
        req.getName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasCreateCollection) {
      val req = bm.getCreateCollection
      val response = if (req.hasOptions) {
        bucket.collections.createCollection(
          req.getScopeName,
          req.getName,
          createCollectionSettings(req.getSettings),
          if (req.getOptions.hasTimeoutMsecs)
            scala.concurrent.duration
              .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
          else DefaultManagementTimeout,
          DefaultRetryStrategy
        )
      } else {
        bucket.collections.createCollection(
          req.getScopeName,
          req.getName,
          createCollectionSettings(req.getSettings)
        )
      }

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasDropCollection) {
      val req = bm.getDropCollection
      val response = if (req.hasOptions) {
        bucket.collections.dropCollection(
          req.getScopeName,
          req.getName,
          if (req.getOptions.hasTimeoutMsecs)
            scala.concurrent.duration
              .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
          else DefaultManagementTimeout,
          DefaultRetryStrategy
        )
      } else {
        bucket.collections.dropCollection(
          req.getScopeName,
          req.getName
        )
      }

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasUpdateCollection) {
      val req = bm.getUpdateCollection
      val response = if (req.hasOptions) {
        bucket.collections.updateCollection(
          req.getScopeName,
          req.getName,
          updateCollectionSettings(req.getSettings),
          if (req.getOptions.hasTimeoutMsecs)
            scala.concurrent.duration
              .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
          else DefaultManagementTimeout,
          DefaultRetryStrategy
        )
      } else {
        bucket.collections.updateCollection(
          req.getScopeName,
          req.getName,
          updateCollectionSettings(req.getSettings)
        )
      }

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else {
      throw new UnsupportedOperationException(new IllegalArgumentException("Unknown operation"))
    }

    result
  }

  private def createCollectionSettings(
      bs: com.couchbase.client.protocol.sdk.bucket.collectionmanager.CreateCollectionSettings
  ): CreateCollectionSettings = {
    var cs = CreateCollectionSettings()
    if (bs.hasExpirySecs) cs = cs.maxExpiry(duration.Duration(bs.getExpirySecs, TimeUnit.SECONDS))
    if (bs.hasHistory) cs = cs.history(bs.getHistory)
    cs
  }

  private def updateCollectionSettings(
      bs: com.couchbase.client.protocol.sdk.bucket.collectionmanager.UpdateCollectionSettings
  ): UpdateCollectionSettings = {
    var cs = UpdateCollectionSettings()
    if (bs.hasExpirySecs) cs = cs.maxExpiry(duration.Duration(bs.getExpirySecs, TimeUnit.SECONDS))
    if (bs.hasHistory) cs = cs.history(bs.getHistory)
    cs
  }

  private def populateResult(
      result: com.couchbase.client.protocol.run.Result.Builder,
      response: Seq[ScopeSpec]
  ): Result.Builder = {
    result.setSdk(
      com.couchbase.client.protocol.sdk.Result
        .newBuilder()
        .setCollectionManagerResult(
          com.couchbase.client.protocol.sdk.bucket.collectionmanager.Result
            .newBuilder()
            .setGetAllScopesResult(
              com.couchbase.client.protocol.sdk.bucket.collectionmanager.GetAllScopesResult.newBuilder
                .addAllResult(response.map(v => populateScopeSpec(v)).asJava)
            )
        )
    )
  }

  private def populateScopeSpec(
      in: ScopeSpec
  ): com.couchbase.client.protocol.sdk.bucket.collectionmanager.ScopeSpec = {
    com.couchbase.client.protocol.sdk.bucket.collectionmanager.ScopeSpec
      .newBuilder()
      .setName(in.name)
      .addAllCollections(in.collections.map(v => populateCollectionSpec(v)).asJava)
      .build
  }

  private def populateCollectionSpec(
      in: CollectionSpec
  ): com.couchbase.client.protocol.sdk.bucket.collectionmanager.CollectionSpec = {
    val builder =
      com.couchbase.client.protocol.sdk.bucket.collectionmanager.CollectionSpec.newBuilder()

    builder
      .setName(in.name)
      .setScopeName(in.scopeName)

    in.expiry.foreach(v => builder.setExpirySecs(v.toSeconds.toInt))
    in.history.foreach(v => builder.setHistory(v))

    builder.build
  }
}
